sympy/physics/units/tests/test_quantities.py

```diff
@@ -573,3 +575,15 @@ def test_physics_constant():
 
     assert not meter.is_physical_constant
     assert not joule.is_physical_constant
+
+
+def test_quantity_substitution():
+    # Define symbol m and quantity expressions
+    m = symbols('m')
+    h = 10 * u.m
+    U = m * u.acceleration_due_to_gravity * h
+    # Substitute m with 10 * u.kg
+    result = U.subs(m, 10 * u.kg)
+    # latex output is as expected
+    expected_latex = '100 \\text{g} \\text{kg} \\text{m}'
+    assert latex(result) == expected_latex, f"Expected: {expected_latex}, got: {latex(result)}"
```

Comment:
The case that I mentioned, replacing m with x should be added since that is a case that we want to watch out for (and it wasn't failing with the code you had before).

Reply:
if I understand what you mean? I should revert back to this line of code 

`if isinstance(new, Quantity) and self != old or isinstance(self, Quantity):`

and add a case where m is replaced with x.